packages/venia-ui/lib/components/Wishlist/GalleryButton/galleryButton.js

```diff
@@ -0,0 +1,83 @@
+import React, { Fragment, useEffect } from 'react';
+import { AlertCircle, Check, Heart, Info } from 'react-feather';
+import { useIntl } from 'react-intl';
+import { useToasts } from '@magento/peregrine';
+import mergeClasses from '@magento/peregrine/lib/util/shallowMerge';
+import { useGalleryButton } from '@magento/peregrine/lib/talons/Wishlist/GalleryButton/useGalleryButton';
+
+import Icon from '../../Icon';
+import defaultClasses from './galleryButton.css';
+import WishlistDialog from '../WishlistDialog';
+
+const CheckIcon = <Icon size={20} src={Check} />;
+const ErrorIcon = <Icon size={20} src={AlertCircle} />;
+const HeartIcon = <Icon size={20} src={Heart} />;
+const InfoIcon = <Icon size={20} src={Info} />;
+
+const GalleryButton = props => {
+    const talonProps = useGalleryButton(props);
+    const {
+        getErrorToastProps,
+        getModalProps,
+        getSuccessToastProps,
+        handleClick,
+        isLoading,
+        isSelected,
+        showLoginToast
+    } = talonProps;
+
+    const classes = mergeClasses(defaultClasses, props.classes);
+
+    const { formatMessage } = useIntl();
+    const [, { addToast }] = useToasts();
+
+    useEffect(() => {
+        if (showLoginToast) {
+            addToast({
+                type: 'info',
+                icon: InfoIcon,
+                message: formatMessage({
+                    id: 'wishlist.galleryButton.loginMessage',
+                    defaultMessage:
+                        'Please sign-in to your Account to save items for later.'
+                }),
+                timeout: 5000
+            });
+        }
+    }, [addToast, formatMessage, showLoginToast]);
+
+    useEffect(() => {
+        if (getSuccessToastProps) {
+            addToast(getSuccessToastProps({ icon: CheckIcon }));
+        }
+    }, [addToast, formatMessage, getSuccessToastProps]);
+
+    useEffect(() => {
+        if (getErrorToastProps) {
+            addToast(getErrorToastProps({ icon: ErrorIcon }));
```

Comment:
Feels like the first time I'm seeing this pattern with toast/props. Specific reason for not just passing them here in the component?

Reply:
This is a pattern widely used in Spectrum and nicely documented by Kent Dodds (who also claims react core uses it a ton). This example is a bit contrived and likely isn't super useful here, but `getModalProps` might show off usage a little better. Can help limit talon response size and provide inversion of control to the component to pass and override whatever props it wants. There's some additional extensibility I haven't added here since I was waiting to discuss with the team, but we can also make event handlers behave as a composite and allow our handler and handlers passed by the component to fire together.

Reply:
Found his blog: https://kentcdodds.com/blog/how-to-give-rendering-control-to-users-with-prop-getters